File: 883-car-fleet/car-fleet.java
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int n = position.length;

        double[][] pair = new double[n][2];
        for (int i = 0; i < n; i++) {
            pair[i][0] = position[i];
            pair[i][1] = speed[i];
        }

        Arrays.sort(pair, (a, b) -> Double.compare(b[0], a[0]));

        Stack<Double> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            double time = (target - pair[i][0]) / pair[i][1];
            stack.push(time);
            if (stack.size() >= 2 && stack.peek() <= stack.get(stack.size() - 2)) {
                stack.pop();
            }
        }

        return stack.size();
    }
}

File: 883-car-fleet/car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        # time: O(n)
        # space: O(1)

        fleets = 0
        current_fleet_time = 0
        for pos, s in sorted(zip(position, speed), key=lambda x: x[0], reverse=True):
            time = (target - pos) / s
            if time > current_fleet_time:
                fleets += 1
                current_fleet_time = time

        return fleets            
        